esp_commands/esp_commands.c

```diff
@@ -0,0 +1,455 @@
+#include "esp_commands.h"
+#include "esp_commands_helpers.h"
+#include "esp_heap_caps.h"
+#include "esp_err.h"
+
+/* Default foreground color */
+#define ANSI_COLOR_DEFAULT 39
+
+/* Pointers to the first and last command in the dedicated section.
+ * See linker.lf for detailed information about the section */
+extern esp_command_t _esp_commands_start;
+extern esp_command_t _esp_commands_end;
+
+/**
+ * @brief Array of pointers to command defining
+ * a set of command. Created while calling
+ * esp_commands_create_set.
+ */
+typedef struct esp_command_set {
+    esp_command_t **cmd_ptr_set;
+    size_t cmd_set_size;
+} esp_command_set_t;
+
+/**
+ * @brief go through all commands registered in the
+ * memory section starting at _esp_commands_start
+ * and ending at _esp_commands_end OR go through all
+ * the commands listed in cmd_set if not NULL
+ */
+#define FOR_EACH_COMMAND(cmd_set, cmd)                                              \
+    for (size_t _i = 0;                                                             \
+         ((cmd_set) == NULL                                                         \
+              ? (((cmd) = &_esp_commands_start + _i),                               \
+                 (&_esp_commands_start + _i) < &_esp_commands_end)                  \
+              : (((cmd) = (cmd_set)->cmd_ptr_set[_i]),                              \
+                 _i < (cmd_set)->cmd_set_size));                                    \
+         ++_i)
+
+/**
+ * @brief returns the number of commands registered
+ * in the .esp_commands section
+ */
+#define ESP_COMMANDS_COUNT (size_t)(&_esp_commands_end - &_esp_commands_start)
+
+/**
+ * @brief find a command by its name in the list of registered
+ * commands or in a command set if the parameter cmd_set is set.
+ *
+ * @note If cmd_set is set to NULL by the caller, then the function
+ * will try to find a command by name from the list of registered
+ * commands in the .esp_commands section
+ *
+ * @param cmd_set command set to find a command from
+ * @param name name of the command to find
+ * @return esp_command_t* pointer to the command matching the command
+ * name, NULL if the command is not found
+ */
+static esp_command_t *find_command_by_name(esp_command_set_t *cmd_set, const char *name)
+{
+    /* no need to check that cmd_set is NULL, if it is, then FOR_EACH_COMMAND will go through
+     * the commands registered in the .esp_commands section */
+    if (!name) {
+        return NULL;
+    }
+    esp_command_t *cmd = NULL;
+    FOR_EACH_COMMAND(cmd_set, cmd) {
+        if (strcmp(cmd->name, name) == 0) {
+            return cmd;
+        }
+    }
+    return NULL;
+}
+
+/** run-time configuration options */
+static esp_commands_config_t s_config = {
+    .heap_alloc_caps = MALLOC_CAP_DEFAULT
+};
+
+/** temporary buffer used for command line parsing */
+static char *s_tmp_line_buf;
+
+esp_err_t esp_commands_init(const esp_commands_config_t *config)
```

Comment:
With the current command implementation, it is not possible to run multiple consoles in parallel is it? (e.g. s_tmp_line_buf is shared)

Is allowing this the plan? I thought we wanted to be able to run multiple repls or similar implementions in parallel (hence the changes to make linenoise capable of this)